### DISCUSSION TITLE
animation in imagemagick - moving and resizing portion of an image to center, and back - how?

### DISCUSSION THREAD

**Discussion by bruce-simonson** [ImageMagick/ImageMagick#2144] [Help] [answered]
Returning to Imagemagick after a hiatus.

Version: ImageMagick 7.0.10-19 Q16 x64 2020-06-12, on Windows 10

I would like to animate the movement of a portion of an image to the center of the image, while leaving the rest of the image the same, hold the moved portion in the center for a while, and then move it back, replacing the portions of the image that were affected by the movement to their original conditions.

There is an animation example that shows how to do this, using -morph, but in the example, the image has to start from the upper left hand corner. The description says that to move from somewhere else, advanced techniques would be needed. I'm hoping someone can help with this.  The animation example is about 2/3rds down this page, under "resize morphing":
  
http://www.imagemagick.org/Usage/anim_mods/#repage_relative
   
```
  convert rose: medical.gif -morph 10 \
          -layers TrimBounds -set dispose previous -coalesce \
          -background black -alpha remove \
          -set delay '%[fx:(t>0&&t<n-1)?10:60]' \
          -duplicate 1,-2-1  -loop 0  morph_resize.gif
```

I would like to be able to specify the portion of the original image that I would like to move (a rectangle, specified by upper left corner in the image, and its dimensions in x and y), and a target location (a target rectangle, specified by a different upper left corner, and size), and animate the movement and resizing of the portion of the image as it moves to the new location.  I would like the image to hold at this location, and then reverse the animation, so that the original image is restored.

To give a concrete example (and what is motivating this), think of a zoom "brady bunch" layout (screenshot), with, say, 5 rows of 6 participants.  After the screenshot is saved, and during post-production, I would like to animate the movement of say, the participant in the second row, 1st column, to the center of the entire image, along with a change in size to say, a height of three rows, and a corresponding width (i.e., without changing the aspect ratio).

The idea is to create an animation that brings focus to this participant, by bringing his or her portion of the screenshot to the center, resizing as it goes.  I want this to be animated.

From:
  
```
+------+------+------+------+------+------+  
|      |      |      |      |      |      |  
+------+------+------+------+------+------+  
|  XX  |      |      |      |      |      |  
+------+------+------+------+------+------+  
|      |      |      |      |      |      |  
+------+------+------+------+------+------+  
|      |      |      |      |      |      |  
+------+------+------+------+------+------+  
|      |      |      |      |      |      |  
+------+------+------+------+------+------+  
```  
to, "slowly" animating the moving rectangle as it moves to the center: 
```  
+------+------+------+------+------+------+  
|      |      |      |      |      |      |  
+------+--+=====================+--+------+  
|      |  |                     |  |      |  
+------+--|    X   X    X   X   |--+------+  
|      |  |      X        X     |  |      |  
+------+--|    X   X    X   X   |--+------+  
|      |  |                     |  |      |  
+------+--+=====================+--+------+  
|      |      |      |      |      |      |  
+------+------+------+------+------+------+  
```  
and, slowly animating the move back.
  
Image example, made up from an internet image (I would like to be able to reverse the animation as well, bringing the relocated central image back to its original location):

Can this animation be done in Imagemagick?

  
[![example of before and after, but want animation showing transition][1]][1]
  

  [1]: https://i.stack.imgur.com/Tw383.png

**Comment by snibgo** [comment, marked as answer]
What version of IM, on what platform?

IM can create the individual frames. Then IM can glue the frames together into a GIF animation, or you an use ffmpeg to make whatever movie format you want.

You want to crop a portion of the image. Then resize that crop, and composite it over the original. The resizing and new position (the "translation") will be animated. For the resize and translate, I suggest you use "-distort SRT", to get smooth movements.

I suggest you have a clock parameter that changes from 0.0 to 1.0 over so many frames. At t=0.0 resize is 1.0 and translation is (0,0). At t=1.0, resize and translation are the maximum values. Intermediate values of t give intermediate resize and translation.

All the IM work can probably be done in a single command, if you have enough memory for all the frames.

See also my page [url=http://im.snibgo.com/animsrt.htm]Animation with SRT[/url]

**Reply by bruce-simonson** [reply]
Version: ImageMagick 7.0.10-19 Q16 x64 2020-06-12, on Windows 10.

Sorry I neglected to include this on the original post, I've updated it with an edit.

Thank you for the link to your page on animation; it looks _very_ promising.

**Comment by snibgo** [comment]
So, implementing your task, suppose we want 100 frames, numbered 0 to 99. Suppose the face we want to crop is width cw, height ch, and the top-left corner is at (cx,cy). So the centre of that crop starts at (x0=cx+cw/2,y0=cy+ch/2). Suppose we want to move the centre of the crop to (x1,y1). The initial resize factor is 1.0 of course, and suppose we want it to end up at factor 2.5.

We make the crop, then loop through the 100 frames. The "+distort SRT" first moves the pixel that was at at (OX,OY) in the crop to (0,0). Then it resizes by a factor DR. Then it moves what was at (0,0) to (NX,NY).

DR, NX and NY are linear interpolations.

As DR has linear variation, it starts with a greater multiplier than it ends with. My linked page shows how to get a constant ratio between frames.

Windows BAT syntax:

```
setlocal enabledelayedexpansion

set nFrames=100

set /A lastFrm=%nFrames%-1

set cw=99
set ch=145
set cx=272
set cy=157

set /A x0=%cx%+%cw%/2
set /A y0=%cy%+%ch%/2

:: New coords for centre of the cropped image:
set x1=509
set y1=447

set ResMin=1.0
set ResMax=2.5

set SRC=..\Tw383.png

%IMG7%magick ^
  %SRC% ^
  -crop %cw%x%ch%+%cx%+%cy% +repage ^
  twcrop.miff

set /A "OX=%cw%/2"
set /A "OY=%ch%/2"

for /L %%I in (0,1,%lastFrm%) do (
  set LZ=000000%%I
  set LZ=!LZ:~-6!

  set "DR=%%[fx:%ResMin%+(%ResMax%-%ResMin%)*%%I/%lastFrm%]"
  set "NX=%%[fx:%x0%+(%x1%-%x0%)*%%I/%lastFrm%]"
  set "NY=%%[fx:%y0%+(%y1%-%y0%)*%%I/%lastFrm%]"

  %IMG7%magick ^
    %SRC% ^
    ^( twcrop.miff ^
       +distort SRT !OX!,!OY!,!DR!,0,!NX!,!NY! ^
       +write x_!LZ!.png ^
    ^) ^
    -background None ^
    -compose Over -layers Flatten ^
    tw283_!LZ!.png
)

:: The frames are tw_283_*.png.
:: Crop them to keep the gif fairly small.

%IMG7%magick ^
  tw283_*.png ^
  -crop 782x776+0+0 +repage ^
  tw283.gif

```
This does the first half. We can repeat all the frames in reverse by modifying the final command:

```
%IMG7%magick ^
  tw283_*.png ^
  -crop 782x776+0+0 +repage ^
  ( -clone 0--1 -reverse ) ^
  tw283b.gif
```

**Comment by GeeMack** [comment]
I see this question has an accepted answer, but as there are often several approaches to a problem when using ImageMagick, I thought I'd add this idea since it's pretty simple and compact. It takes any size input image, and you don't even have to know any actual dimensions or offsets.

This ImageMagick v7 and Windows BAT script will produce a GIF with a selected sub-image zooming larger while moving to the center, pausing a moment, then receding back to its original size and location. The only required variables are the input image, IMAGE, the number of columns and rows in the grid, WIDE and HIGH, and the cell coordinates of the cell to use, OVER and DOWN. Virtually all the calculations are done within the ImageMagick command.

     @echo off
     :: =====================================================
     
     set IMAGE=allfaces.png
     
     set WIDE=6
     set HIGH=5
     
     set OVER=3
     set DOWN=1
     
     set /a INUMBER=(WIDE*(DOWN-1))+OVER
     
     magick %IMAGE% -write mpr:img0 -background none ^
        -crop %WIDE%x%HIGH%@ -delete %INUMBER%--1 -reverse -delete 1--1 ^
        -duplicate 11 -resize "%%[fx:t*200/(n-1^)+100]%%" ^
        -set page "%%[fx:u.page.width]x%%[fx:u.page.height]+%%[fx:u.page.x]+%%[fx:u.page.y]" ^
        -extent ^"%%[page]^
           -%%[fx:(t*(((u[-1].page.width-u[-1].w^)/2^)-u[-1].page.x^)/(n-1^)^)+u[-1].page.x] ^
           -%%[fx:(t*(((u[-1].page.height-u[-1].h^)/2^)-u[-1].page.y^)/(n-1^)^)+u[-1].page.y] ^" ^
        +repage null: mpr:img0 -compose dstover -layers composite ^
        -set delay %%[fx:t==(n-1)?100:5] ( -clone 0--2 -reverse ^) ^
        -loop 0 animateface.gif
     
     exit /b
     :: =====================================================

The IM command starts by reading in the original image and cropping it into individual cells. Then it deletes all the cells we're not using. It duplicates the remaining cell the number of steps we want the animation to zoom. I used "-duplicate 11" to make 12 steps, but more steps make a smoother animation and larger file.

Then it resizes them all incrementally starting from the original size and going to 300% of that. After setting some paging geometry it uses "-extent" to resize all the canvases back to the input image dimensions *and* locate each image in equal steps from their starting location to fully centered.

After that it brings back a copy of the input image "mpr:img0", and composites it as a background behind all the resized ones. The delay is set to 100 on the largest zoom to make the animation pause there, and to 5 on the rest to make a smooth zoom. Next, all but the last image are copied, the order reversed, and they're added to the end of the current list.

The output is an animated GIF with the selected cell zooming from its starting place to the center and back.
